File: sdk_samples/samples/C++/07_multistage_anpr/07_multistage_anpr.cpp
/**
 * CARMEN Video SDK
 *
 * @category    C++ Sample
 * @package     CARMEN Video C++ Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * uses multistage ANPR.
 */

#include <iostream>

#include "../common/utils.hpp"

#include <carmen/AnprBuilder.hpp>
#include <carmen/MmrBuilder.hpp>
#include <carmen/StreamProcessorBuilder.hpp>
#include <carmen/Logger.hpp>


void onEventCallback(const cm::Event& e) {
    try {
        std::cout << "------------------------------------------------------" << std::endl;
        std::cout << "Timestamp: " << e.timestamp() << " ms" << std::endl;

        //Plate
        std::cout << "Plate text: " << e.vehicle().plate().text() << std::endl;
        std::cout << "Country: " << e.vehicle().plate().country() << std::endl;

        //MMR
        std::cout << "Make: " << e.vehicle().attributes().make() << std::endl;
        std::cout << "Model: " << e.vehicle().attributes().model() << std::endl;
        std::cout << "Color: " << colorRGBIntToString(e.vehicle().attributes().color()) << std::endl;
        std::cout << "Category: " << e.vehicle().attributes().category() << std::endl;

        std::cout << std::endl;

    } catch(const std::exception& e) {
        std::cout << "Exception in event callback " << e.what() << std::endl;
    } catch(...) {
        std::cout << "Exception in event callback" << std::endl;
    }
}


std::optional<int> anpr_stage_callback(const cm::anpr::StageCallbackParam& info) {
    std::cout << "stage callback" << std::endl;
    for(const auto& stage : info.stages) {
        std::cout << stage.plateTextUtf8 << " - " << stage.confidence << std::endl;
    }

    auto lastIndex = info.stages.size() - 1;
    auto& lastStage = info.stages[lastIndex];
    if(!lastStage.plateTextUtf8.empty() && ((lastStage.confidence>50) || info.wasLastStage)) {
        return lastIndex;
    }
    return std::nullopt;
}


void add_anpr(cm::video::StreamProcessorBuilder& builder, const char* stage1_region, const char* stage2_region) {
    cm::anpr::Anpr anpr = cm::anpr::AnprBuilder()
            .type(cm::anpr::AnprBuilder::Type::LOCAL) // use LOCAL_GO for CARMEN_GO engines
            .localConcurrencyLimit(1)
            .build();

    cm::anpr::AnprProfile anprProfile;
    anprProfile.stageCallback = anpr_stage_callback;
    anprProfile.stages.push_back(cm::anpr::AnprStage{stage1_region, std::nullopt, std::nullopt, {}});
    anprProfile.stages.push_back(cm::anpr::AnprStage{stage2_region, std::nullopt, std::nullopt, {}});
    auto anprProfileId = anpr.registerProfile(anprProfile);

    builder.anpr(anpr).anprProfileId(anprProfileId);
}


void add_mmr(cm::video::StreamProcessorBuilder& builder, const char* mmrRegion) {
    cm::mmr::Mmr mmr = cm::mmr::MmrBuilder()
            .type(cm::mmr::MmrBuilder::Type::LOCAL)
            .build();
    cm::mmr::MmrProfile mmrProfile{mmrRegion, std::nullopt};
    auto mmrProfileId = mmr.registerProfile(mmrProfile);
    builder.mmr(mmr);
    builder.mmrProfileId(mmrProfileId);
}


cm::video::StreamProcessor buildStreamProcessor(
        const char* stage1_region,
        const char* stage2_region,
        const char* streamUrl,
        const char* mmrRegion) {

    cm::video::StreamProcessorBuilder streamProcessorBuilder;
    streamProcessorBuilder
        .source(streamUrl)
        .name("Stream 1")
        .eventCallback(onEventCallback)
        .statusChangeCallback(commonStatusCallback)
        .roi({{0.0,  0.0},
              {0.95, 0.0},
              {0.95, 0.95},
              {0.0,  0.95}})
        .autoReconnect(true);

    add_anpr(streamProcessorBuilder, stage1_region, stage2_region);

    if(mmrRegion != nullptr && mmrRegion[0] != 0) {
        add_mmr(streamProcessorBuilder, mmrRegion);
    }

    return streamProcessorBuilder.build();
}


int main(int argc, const char** argv) {
    try {
        if(argc < 3) {
            std::cout << "Usage: " << argv[0] << "<region code 1st> <region code 2nd> <stream url / video file> [mmr-region]\n" << std::endl;

            //  Region code:
            //      See in Reference Manual: Region List
            //      For "mmr-region" you have to include the "mmr-" prefix like it appears in gxsd.dat (e.g.: "mmr-eur")
            //
            //  Stream url examples:
            //      "rtsp://username:password@192.168.1.2:8994"
            //      "http://192.168.1.2:9901/video.mjpeg"
            //  Video file example:
            //      "file:///C:\\video.mp4"

            return -1;
        }

        const char* stage1_region = argv[1];
        const char* stage2_region = argv[2];
        const char* streamUrl = argv[3];
        const char* mmrRegion = nullptr;
        if(argc >= 5) {
            mmrRegion = argv[4];
        }

        cm::log::GlobalLogger::setMinLevel(cm::log::LogLevel::WARNING);

        cm::video::StreamProcessor streamProcessor = buildStreamProcessor(stage1_region, stage2_region, streamUrl, mmrRegion);

        std::cout << "Please press 'q' then 'Enter' to stop stream processing." << std::endl;

        streamProcessor.start();

        while(std::cin.get() != 'q');

        streamProcessor.stop();

    } catch(const std::exception& e) {
        std::cout << e.what() << std::endl;
        return -1;
    }
    return 0;
}

File: sdk_samples/samples/CSharp/07_multistage_anpr/07_multistage_anpr.cs
/**
 * CARMEN Video SDK
 *
 * @category    CSharp Sample
 * @package     CARMEN Video CSharp Sample
 *
 * @copyright   2024 Adaptive Recognition
 * @licence     https://adaptiverecognition.com/eula/
 *
 * This sample shows how you can set up a process with CARMEN Video SDK that
 * uses multistage ANPR.
 */

using Carmen;
using Carmen.Log;
using Carmen.Anpr;
using Carmen.Mmr;
using Carmen.Video;

class Sample07_multistage_anpr
{
    static void EventHandlerCallback(Event e)
    {
        try
        {
            using (e)
            {
                Console.WriteLine("------------------------------------------------------");
                Console.WriteLine("Unix timestamp: " + e.Timestamp + " ms (" + e.DateTime.ToLocalTime() + ")");
                Console.WriteLine();

                //Plate
                Console.WriteLine("Plate text: " + e.Vehicle.Plate.Text);
                Console.WriteLine("Country: " + e.Vehicle.Plate.Country);

                //MMR
                Console.WriteLine("Make: " + e.Vehicle.Attributes.Make);
                Console.WriteLine("Model: " + e.Vehicle.Attributes.Model);
                Console.WriteLine("Color: " + e.Vehicle.Attributes.Color);
                Console.WriteLine("Category: " + e.Vehicle.Attributes.Category);

                Console.WriteLine();
            }
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("Exception in event callback. " + ex);
        }
    }

    static int? AnprStageCallback(Carmen.Anpr.StageCallbackParam info)
    {
        try
        {
            Console.WriteLine("stage callback");
            foreach (var stage in info.Stages)
            {
                Console.WriteLine(stage.PlateTextUtf8 + " - " + stage.Confidence);
            }

            int lastIndex = info.Stages.Count - 1;
            var lastStage = info.Stages[lastIndex];
            if (!string.IsNullOrEmpty(lastStage.PlateTextUtf8) && ((lastStage.Confidence > 50) || info.WasLastStage))
            {
                return lastIndex;
            }
            return null;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine("Exception in AnprStageCallback callback. " + ex);
            return null;
        }
    }

    static void AddAnpr(StreamProcessor.StreamProcessorBuilder builder, string stage1Region, string stage2Region)
    {
        using Anpr.AnprBuilder anprBuilder = Anpr.Builder();
        using Anpr anpr = anprBuilder
            .Type(AnprType.Local) // use LocalGo for CARMEN_GO engines
            .LocalConcurrencyLimit(1)
            .Build();

        AnprProfile anprProfile = new AnprProfile();
        anprProfile.StageCallback = AnprStageCallback;
        var stage1 = new AnprStage(stage1Region, null, null, new Dictionary<string, string>());
        var stage2 = new AnprStage(stage2Region);
        anprProfile.Stages.Add(stage1);
        anprProfile.Stages.Add(stage2);
        AnprProfileId anprProfileId = anpr.RegisterProfile(anprProfile);

        builder.Anpr(anpr);
        builder.AnprProfileId(anprProfileId);
    }

    static void AddMmr(StreamProcessor.StreamProcessorBuilder builder, string mmrRegion)
    {
        using Mmr.MmrBuilder mmrBuilder = Mmr.Builder();
        using Mmr mmr = mmrBuilder
            .Type(MmrType.Local)
            .Build();

        MmrProfile mmrProfile = new MmrProfile(mmrRegion, null);
        var mmrProfileId = mmr.RegisterProfile(mmrProfile);

        builder.Mmr(mmr);
        builder.MmrProfileId(mmrProfileId);
    }

    static StreamProcessor BuildStreamProcessor(String stage1Region, String stage2Region, String streamUrl, String? mmrRegion)
    {
        using StreamProcessor.StreamProcessorBuilder streamProcessorBuilder = StreamProcessor.Builder();
        streamProcessorBuilder
            .Source(streamUrl)
            .Name("Stream 1")
            .EventCallback(EventHandlerCallback)
            .StatusChangeCallback
            (
                (stream1, status) => { Utils.CommonStatusCallback(stream1, status); }
            )
            .Roi(new List<Carmen.Point>
                {
                    new Carmen.Point(0.0, 0.0),
                    new Carmen.Point(0.95, 0.0),
                    new Carmen.Point(0.95, 0.95),
                    new Carmen.Point(0.0, 0.95)
                }
            )
            .AutoReconnect(true);

        AddAnpr(streamProcessorBuilder, stage1Region, stage2Region);

        if (!string.IsNullOrEmpty(mmrRegion))
        {
            AddMmr(streamProcessorBuilder, mmrRegion);
        }

        return streamProcessorBuilder.Build();
    }


    static void Main(string[] args)
    {
        try
        {
            if (args.Length < 3)
            {
                Console.WriteLine(
                    "Usage: <program name> <region code 1st> <region code 2nd> <stream url / video file> [mmr-region]");

                //  Region code:
                //      See in Reference Manual: Region List
                //
                //  Stream url examples:
                //      "rtsp://username:password@192.168.1.2:8994"
                //      "http://192.168.1.2:9901/video.mjpeg"
                //  Video file example:
                //      "file:C:/video.mp4"

                Console.ReadKey();
                return;
            }

            String stage1Region = args[0];
            String stage2Region = args[1];
            String streamUrl = args[2];
            String? mmrRegion = args.Length > 3 ? args[3] : null;

            GlobalLogger.SetMinLevel(LogLevel.Warning);

            using StreamProcessor stream = BuildStreamProcessor(stage1Region, stage2Region, streamUrl, mmrRegion);

            Console.WriteLine("Please press 'q' then 'Enter' to stop stream processing.");

            stream.Start();

            while (Console.ReadLine() != "q") ;

            stream.Stop();
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }

        Console.WriteLine("Please press a key to exit the program.");
        Console.ReadKey();
    }
}